### ISSUE TITLE
Attestations Overwriting Identity

### ISSUE THREAD

**Issue by micahalcorn** [OriginProtocol/origin#2728]
In random instances, doing an attestation (with Twitter, for example) overwrites other parts of a user's profile. I lost my name, picture, email and phone on Mainnet when connecting Twitter and @tomlinton had a similar experience losing his phone after adding email.

@franckc

**Comment by micahalcorn** [comment]
Resolved 🤞

**Comment by tomlinton** [comment]
I lost email and phone after adding Twitter. I grabbed this screenshot:

<img width="787" alt="Screen Shot 2019-07-24 at 2 43 17 PM" src="https://user-images.githubusercontent.com/489202/61761195-0ce58400-ae22-11e9-9d7b-f4dbe60639ca.png">

The `0x488...` account is the proxy, and `0xd15...` is my actual account. So it looks like both the signatures of the attestations failed to validate. This would probably mean they wouldn't get sent with the next identity object and so it would like like the Twitter attestation overwrites them.

Below is the identity after this overwrite:

```
{
  "data": {
    "web3": {
      "account": {
        "id": "0xd15f105E2667Fd52313877585D4a0b316C7D8a20",
        "owner": {
          "id": "0xd15f105E2667Fd52313877585D4a0b316C7D8a20"
        },
        "proxy": {
          "id": "0x488C384d2FDddb08ECE822fae535CA8AF0e9D4Cb"
        },
        "identity": {
          "id": "0xd15f105E2667Fd52313877585D4a0b316C7D8a20",
          "firstName": "Tom",
          "lastName": "Linton",
          "fullName": "Tom Linton",
          "description": "Engineer at Origin Protocol",
          "avatarUrl": "ipfs://QmaKc5QncA76mBB2WadbGzi3UN9UeEK2sKjC1u8XyJuMJQ",
          "avatarUrlExpanded": "https://ipfs.originprotocol.com/ipfs/QmaKc5QncA76mBB2WadbGzi3UN9UeEK2sKjC1u8XyJuMJQ",
          "strength": 40,
          "attestations": [
            "{\"schemaId\":\"https://schema.originprotocol.com/attestation_1.0.0.json\",\"data\":{\"issuer\":{\"name\":\"Origin Protocol\",\"url\":\"https://www.originprotocol.com\",\"ethAddress\":\"0x8EAbA82d8D1046E4F242D4501aeBB1a6d4b5C4Aa\"},\"issueDate\":\"2019-07-24T02:42:48.708Z\",\"attestation\":{\"verificationMethod\":{\"oAuth\":true},\"site\":{\"siteName\":\"twitter.com\",\"userId\":{\"raw\":\"44052572\"},\"username\":{\"raw\":\"tmlntn\"},\"profileUrl\":{\"raw\":\"https://twitter.com/tmlntn\"}}}},\"signature\":{\"bytes\":\"0x5ed330969c778d9d592ee812fbab5266c8fee72f120287c7b41497a1b4b028f50d4a4557065ccddbfa8b57110d73779c6ba70b2a59fab51eed45661df499c3a61b\",\"version\":\"1.0.0\"}}"
          ],
          "verifiedAttestations": [
            {
              "id": "twitter",
              "rawData": "{\"schemaId\":\"https://schema.originprotocol.com/attestation_1.0.0.json\",\"data\":{\"issuer\":{\"name\":\"Origin Protocol\",\"url\":\"https://www.originprotocol.com\",\"ethAddress\":\"0x8EAbA82d8D1046E4F242D4501aeBB1a6d4b5C4Aa\"},\"issueDate\":\"2019-07-24T02:42:48.708Z\",\"attestation\":{\"verificationMethod\":{\"oAuth\":true},\"site\":{\"siteName\":\"twitter.com\",\"userId\":{\"raw\":\"44052572\"},\"username\":{\"raw\":\"tmlntn\"},\"profileUrl\":{\"raw\":\"https://twitter.com/tmlntn\"}}}},\"signature\":{\"bytes\":\"0x5ed330969c778d9d592ee812fbab5266c8fee72f120287c7b41497a1b4b028f50d4a4557065ccddbfa8b57110d73779c6ba70b2a59fab51eed45661df499c3a61b\",\"version\":\"1.0.0\"}}",
              "properties": [
                {
                  "type": "created",
                  "value": "2019-07-24T02:42:48.708Z"
                },
                {
                  "type": "userId",
                  "value": "44052572"
                },
                {
                  "type": "username",
                  "value": "tmlntn"
                },
                {
                  "type": "profileUrl",
                  "value": "https://twitter.com/tmlntn"
                }
              ]
            }
          ]
        }
      }
    }
  }
}```

**Comment by shahthepro** [comment]
@tomlinton Was there any change in this env key `ATTESTATION_SIGNING_KEY`? From what I get from the screenshot, It seems like the attestation data was signed by accounts starting with `0xB2e...` and `0xAa0...` while it should have been signed by `0x8EAbA82d8D1046E4F242D4501aeBB1a6d4b5C4Aa`

**Comment by shahthepro** [comment]
On a closer look, we hash the user address and attestation data and then sign it with `ATTESTATION_SIGNING_KEY`.

If we hash a different user address and the attestation data, we get some random address as signer while doing a `ecrecover`. This is why we are getting signer addresses as `0xB2e...` and `0xAa0`.

Probably the accounts [returned by these lines](https://github.com/OriginProtocol/origin/blob/master/packages/graphql/src/mutations/identity/deployIdentity.js#L24-L27) are incorrect.

**Comment by tomlinton** [comment]
It won't be `ATTESTATION_SIGNING_KEY`. I reproduced this a few times and it happened on attestations done within 2-3 minutes of each other.

**Comment by franckc** [comment]
Couple quick thoughts to help debug this issue:
 - [hasProxyRaw](https://github.com/OriginProtocol/origin/blob/master/packages/graphql/src/utils/proxy.js#L52) and [proxyOwnerRaw](https://github.com/OriginProtocol/origin/blob/master/packages/graphql/src/utils/proxy.js#L69) eat exceptions silently. How about logging those errors in Sentry ?
 - I wonder if the fact that those methods returns are [cached](https://github.com/OriginProtocol/origin/blob/master/packages/graphql/src/utils/proxy.js#L73) could cause issue in some cases ? Now disabling the caching could cause performance regression so we probably want to be careful... But maybe we should invalidate the cache for example when a user creates a proxy ?

**Comment by shahthepro** [comment]
@franckc 
> * [hasProxyRaw](https://github.com/OriginProtocol/origin/blob/master/packages/graphql/src/utils/proxy.js#L52) and [proxyOwnerRaw](https://github.com/OriginProtocol/origin/blob/master/packages/graphql/src/utils/proxy.js#L69) eat exceptions silently. How about logging those errors in Sentry ?

Yes, we should definitely log that. But from the screenshot that Tom posted, it seems that it returned both Proxy address and owner address correctly.

> * I wonder if the fact that those methods returns are [cached](https://github.com/OriginProtocol/origin/blob/master/packages/graphql/src/utils/proxy.js#L73) could cause issue in some cases ? Now disabling the caching could cause performance regression so we probably want to be careful... But maybe we should invalidate the cache for example when a user creates a proxy ?

We do clear the cache [using this function](https://github.com/OriginProtocol/origin/blob/master/packages/graphql/src/utils/proxy.js#L76-L79), after a proxy has been deployed.

**Comment by shahthepro** [comment]
@micahalcorn @tomlinton 
I tried by creating a [wallet](0xEB7bC1433A6C46104EC414fc0bCB2781545fEF8E) and [proxy address](https://etherscan.io/address/0x465270da7a0957041E9FCEf073dEa59601f54F58) and then added attestations multiple times within a short span of time. [Still couldn't get to reproduce this](https://shoporigin.com/#/user/0x465270da7a0957041E9FCEf073dEa59601f54F58). 

Can anyone help me with reproducing this?

**Comment by tomlinton** [comment]
I think most of the work in solving this issue is finding the reproduction. Only things of note are:

- Account created in mobile
- This error occured in desktop after an import
- The account had made an offer on a listing

**Comment by shahthepro** [comment]
@micahalcorn The issue of profile data (like name, photo, description) and attestations getting lost will be fixed with #2845 

@tomlinton The issue that [you mentioned in this comment](https://github.com/OriginProtocol/origin/issues/2728#issuecomment-514458489) is different one from what @micahalcorn reported. In your case, you didn't lose profile data or attestation at the DApp level. GraphQL failed to validate the signature. On a closer look it seems that the [`JSON.parse()` that happens here](https://github.com/OriginProtocol/origin/blob/master/packages/graphql/src/mutations/identity/deployIdentity.js#L33) might be the culprit for this one. I suspect it isn't deterministic. So we can get different value while signing and verifying. We can probably move this into a separate issue. I'll have another PR for this.
